Velocity Modifiers

Rayoule opened the pull request:
- This PR adds velocity modifiers to replace drag and acceleration.
- A new modifier "Noise" adds a simple sinusoïdal 2D noise.
- A new example velocity.rs replaces drag.rs by illustrating Constant, Drag and Noise modifiers.
- There is also a new PrecalculatedParticleVariables struct used in systems.rs when applying the modifiers to the particles transforms: It is handful to avoid doing multiple normalize() or length() on velocity when they are used in multiple modifiers.

Let me know what you think of this !

abnormalbrain commented on src/systems.rs:
```diff
@@ -337,18 +335,41 @@ pub(crate) fn particle_transform(
+                        velocity.0 += *v * delta_time;
+                    }
+
+                    ScalarAcceleration(v) => {
```
It might be useful to also have a plain `Acceleration` variant that is a `Vec3`. I think you mentioned it elsewhere, was the main blocker from doing that that we don't have a lerp for `Vec3`?

Rayoule replied:
There is already a Vec3 variant that was poorly named Constant.
I will rename this Vector, and ScalarAcceleration -> Scalar.
There was no blocker, but indeed I mentioned that it would be cool to have key values for Vec3 and f32. Not only lerp, which is cool, but don't offer as much control as key values (like gradient for ColorOverTime).

abnormalbrain commented on src/values.rs:
```diff
@@ -808,3 +808,120 @@ impl Default for SinWave {
+        }
+    }
+
+    /// evaluates the noise at a given position and time
```
```suggestion
    /// Evaluates the noise at a given position and time
```

abnormalbrain commented on src/systems.rs:
```diff
@@ -337,18 +335,41 @@ pub(crate) fn particle_transform(
+                    Noise(n) => {
+                        let offset = n.sample(
+                            Vec2::new(transform.translation.x, transform.translation.y),
+                            time.elapsed_seconds_wrapped(),
```
I think we should make the same distinction here as we do with delta_time above, using `elapsed_seconds_wrapped()` if we're respecting time scaling and `raw_elapsed_seconds_wrapped()` if we're not, just for consistencies sake.

abnormalbrain reviewed (commented):
A few minor comments but overall this looks awesome, thank you!

Rayoule commented:
As we mentioned previously, a VectorOverTime was needed.
And with a VectorOverTime, we want to be able to have a constant, a lerp, but also a gradient-like value/key pair to animate the vector over the particle's life.

So I changed the Gradient struct to a "Curve<T>" struct that is generic and is used as Color, Vec3 and f32. This change allow us to animate Vec3 and f32 with keys, when it was only possible for Color. Plus we can now use a VectorOverTime as a VelocityModifier, its more freedom for the user !

The name "Curve" was inspired by Unreal's "Cascade" particle system editor, that features the same value/key pair animation method, and "Gradient" wasn't relevant for f32 and Vec3 values.

I kept the ValueOverTime/ColorOverTime/VectorOverTime enums implementations separated because they probably won't handle the same variants: like Sin(SinWave) for ValueOverTime for example. Such a variant wouldn't make any sense with Vectors or Colors. Also we might want to rotate vectors, or change HSV for colors, etc... I ended up thinking that these should not be in the same enum generic type despite all having some common variants.

I updated the examples and "fixed" shape_emitter.rs where it was using a Curve<Color> but its lifetime + max_distance didn't allow the particles to finish their lifetime, and by extension their Curve<Color>. I lowered the lifetime, and now we can see them nicely fade.

Let me know what you think !

abnormalbrain commented:
Fantastic!
